packages/go_router/lib/src/builder.dart

```diff
@@ -163,34 +164,69 @@ class RouteBuilder {
 
       keyToPages.putIfAbsent(goRouteNavKey, () => <Page<dynamic>>[]).add(page);
 
-      _buildRecursive(context, matchList, startIndex + 1, pop, routerNeglect,
-          keyToPages, newParams, navigatorKey);
-    } else if (route is ShellRoute) {
+      return _buildRecursive(context, matchList, startIndex + 1, pop,
+              routerNeglect, keyToPages, newParams, navigatorKey) ??
+          state;
+    } else if (route is ShellRouteBase) {
+      if (startIndex + 1 >= matchList.matches.length) {
+        throw _RouteBuilderError('Shell routes must always have child routes');
+      }
+
       // The key for the Navigator that will display this ShellRoute's page.
       final GlobalKey<NavigatorState> parentNavigatorKey = navigatorKey;
 
-      // The key to provide to the ShellRoute's Navigator.
-      final GlobalKey<NavigatorState> shellNavigatorKey = route.navigatorKey;
-
       // Add an entry for the parent navigator if none exists.
       keyToPages.putIfAbsent(parentNavigatorKey, () => <Page<dynamic>>[]);
 
-      // Add an entry for the shell route's navigator
-      keyToPages.putIfAbsent(shellNavigatorKey, () => <Page<dynamic>>[]);
-
       // Calling _buildRecursive can result in adding pages to the
       // parentNavigatorKey entry's list. Store the current length so
       // that the page for this ShellRoute is placed at the right index.
       final int shellPageIdx = keyToPages[parentNavigatorKey]!.length;
 
-      // Build the remaining pages
-      _buildRecursive(context, matchList, startIndex + 1, pop, routerNeglect,
-          keyToPages, newParams, shellNavigatorKey);
+      // Get the current child route of this shell route from the match list.
+      final RouteBase childRoute = matchList.matches[startIndex + 1].route;
+
+      // The key to provide to the shell route's Navigator.
+      final GlobalKey<NavigatorState>? shellNavigatorKey =
+          route.navigatorKeyForChildRoute(childRoute);
+      if (shellNavigatorKey == null) {
+        throw _RouteBuilderError(
+            'Shell routes must always have a navigator key');
+      }
+
+      // Add an entry for the shell route's navigator
+      keyToPages.putIfAbsent(shellNavigatorKey, () => <Page<dynamic>>[]);
+
+      // Build the remaining pages and retrieve the state for the top of the
+      // navigation stack
+      final GoRouterState? topRouterState = _buildRecursive(
+        context,
+        matchList,
+        startIndex + 1,
+        pop,
+        routerNeglect,
+        keyToPages,
+        newParams,
+        shellNavigatorKey,
+      );
 
-      // Build the Navigator
-      final Widget child = _buildNavigator(
+      Widget child = _buildNavigator(
```

Comment:
Setting the `restorationScopeId` for a `GoRouter` object throws a `FlutterError` with diagnostic notes:
```
ErrorSummary('Multiple owners claimed child RestorationBuckets with the same IDs.')
ErrorDescription('The following IDs were claimed multiple times from the parent $this:')
```

Reply:
Good catch! Wondered if this perhaps would be an issue when I looked at the existing implementation, and was meaning to test it (but never got around to it). 

Perhaps we need to add `restorationScopeId` to the shell route classes, or derive the `restorationScopeId` for nested Navigators from the one passed to `GoRouter`. Or both.

Reply:
@chunhtai, do you think we should handle support for `restorationScopeId` in shell routes (this applies to both ShellRoute and StatefulShellRoute) in a separate PR?

Reply:
I think we can just do it here, it is merely passing in a parameter I think?

Reply:
Yes, I guess it's mainly adding a field for it on `ShellRoute` & `ShellRouteBranch` and passing that to the navigator. Question is if we need any validation, i.e. what should happen if `restorationScopeId` is specified for `GoRouter` and not for a shell route, or the other way around? Or is this outside the responsibility of GoRouter?

Reply:
you can verify that the navigator is built with correct restoration id and assert it contains the right one. You can assume the navigator works correctly once the restoration id is assigned

Reply:
👍 Sound good.